File: edition4/js/ch02/2-1.js
function Node(data, next) {
  this.data = data;
  this.next = next;
};

function deleteNode(head, target) {

  if (head === target) {
    // head is target, set head to second element and return
    head = head.next;
  } else {
    // start at the second element
    var curr = head.next;
    var prev = head;
    // prev = curr;
    // curr = curr.next;

    while(curr) {
      if (curr === target) {
        prev.next = curr.next;
      }

      prev = curr;
      curr = curr.next;
    }
  }

  return head;
}

/**
 * Deletes duplicates from list with an additional buffer.
 */
function deleteDups(head) {
  var buffer = {};
  var curr = head;

  while(curr) {
    if (buffer[curr.data]) {
      head = deleteNode(head, curr);
    } else {
      buffer[curr.data] = curr;
    }

    curr = curr.next
  }

  return head;
}

// tests
var assert = require('assert');

describe('deleteNode', function() {

  it('should delete the head', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    head = deleteNode(head, head);
    assert.equal(head.data, n1.data);
  });

  it('should delete the node', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    head = deleteNode(head, n1);
    assert.equal(head.next.data, n2.data);
  });

});

describe('deleteDups', function() {

  it('should not delete anything', function() {
    var n3 = new Node('4', null);
    var n2 = new Node('3', n3);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    assert.deepEqual(deleteDups(head), head);
  });

  it('should delete duplicates', function() {
    var n3 = new Node('3', null);
    var n2 = new Node('2', n3);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    head = deleteDups(head);
    assert.equal(head.data, '1');
    assert.equal(head.next.data, '2');
    assert.equal(head.next.next.data, '3');
  });

  it('should delete everything but head', function() {
    var n3 = new Node('1', null);
    var n2 = new Node('1', n3);
    var n1 = new Node('1', n2);
    var head = new Node('1', n1);
    head = deleteDups(head);
    assert.equal(head.data, '1');
    assert.equal(head.next, null);
  });

});

File: edition4/c/ch02/2-1.cpp
#include <assert.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <map>
#include <string>

using namespace std;

class Node {
  public:
    int value;
    Node *next;
    Node(int value, Node *next);
};

Node::Node(int value, Node *next) {
  this->value = value;
  this->next = next;
}

class LinkedList {
  public:
    Node *head;
    LinkedList();
    void add(int value);
    void removeDups();
    void removeDupsBetter();
    int length();
    void print();
};

LinkedList::LinkedList() {
  this->head = NULL;
}

void LinkedList::add(int value) {
  if (!this->head) {
    this->head = new Node(value, NULL);
  } else {
    Node *temp = this->head;

    while (temp->next) {
      temp = temp->next;
    }

    temp->next = new Node(value, NULL);
  }
}

void LinkedList::removeDups() {
  Node *prev = this->head;
  Node *curr = this->head->next;
  while (curr) {
    Node *runner = this->head;

    while (runner != curr) {
      if (runner->value == curr->value) {
        Node * temp = curr;
        curr = curr->next;
        prev->next = curr;
        delete temp;
        break;
      }

      runner = runner->next;
    }
    if (curr == runner) {
      prev = curr;
      curr = curr->next;
    }
  }
}

void LinkedList::removeDupsBetter() {
  map<int, bool> m;
  Node *prev = this->head;
  Node *curr = this->head->next;

  // seed the map with value of first list element that we are skipping.
  // m.insert(prev->value, true);
  m.insert(pair<int,bool>(prev->value,true));

  while (curr) {
    // check hash, if exists, delete from list
    if (m.find(curr->value) != m.end()) {
        Node * temp = curr;
        curr = curr->next;
        prev->next = curr;
        delete temp;

        continue;
    } else {
      m.insert(pair<int,bool>(curr->value,true));
    }

    // add to map and increment the current pointer
    prev = curr;
    curr = curr->next;
  }
}

int LinkedList::length() {
  int length = 0;
  Node *curr = this->head;

  while (curr) {
    length++;
    curr = curr->next;
  }

  return length;
}

void LinkedList::print() {
  Node *curr = this->head;

  while (curr) {
    cout << curr->value << " -> ";
    curr = curr->next;
  }

  cout << "NULL" << endl;
}

int main() {
  LinkedList list;
  list.add(1);
  list.add(2);
  list.add(3);
  list.removeDupsBetter();
  assert(list.length() == 3);

  list.add(2);
  list.removeDupsBetter();
  assert(list.length() == 3);

  list.add(2);
  list.add(2);
  list.add(2);
  list.add(2);
  list.add(2);
  list.removeDupsBetter();
  assert(list.length() == 3);

  return 0;
}
